java/code/src/com/suse/manager/webui/controllers/clusters/ClustersController.java

```diff
@@ -0,0 +1,595 @@
+        data.put("flashMessage", FlashScopeHelper.flash(request));
+        data.put("contentCluster", GSON.toJson(
+                toClusterResponse(cluster, getClusterProvider(cluster.getProvider()))));
+        MinionController.addActionChains(user, data);
+        return new ModelAndView(data, "controllers/clusters/templates/join.jade");
+    }
+
+    private static ModelAndView showClusterUpgrade(Request request, Response response, User user) {
+        Cluster cluster = getCluster(request);
+        Map<String, Object> data = new HashMap<>();
+        data.put("flashMessage", FlashScopeHelper.flash(request));
+        data.put("contentCluster", GSON.toJson(
+                toClusterResponse(cluster, getClusterProvider(cluster.getProvider()))));
+        MinionController.addActionChains(user, data);
+        return new ModelAndView(data, "controllers/clusters/templates/upgrade.jade");
+    }
+
+    private static String addCluster(Request request, Response response, User user) {
+        Optional<Map<String, Object>> json = parseJson(request, response);
+        if (json.isEmpty()) {
+            return json(response, HttpStatus.SC_BAD_REQUEST,
+                    ResultJson.error("request_error"));
+        }
+
+        // TODO validate input
+        String name = (String)json.get().get("name");
+        String label = (String)json.get().get("label");
+        String description = (String)json.get().get("description");
+        String provider = (String)json.get().get("provider");
+        long managementNodeId = ((Number)json.get().get("managementNodeId")).longValue();
+        Map<String, Object> managementSettings = (Map<String, Object>)json.get().get("managementSettings");
+        Cluster cluster;
+        try {
+            cluster = clusterManager.addCluster(name, label, description,
+                    managementNodeId, provider, managementSettings, user);
+            HibernateFactory.getSession().getTransaction().commit();
+        }
+        catch (Exception e) {
+            LOG.error("Adding cluster failed", e);
+            return json(response, HttpStatus.SC_INTERNAL_SERVER_ERROR, ResultJson.error(e.getMessage()));
+        }
+        FlashScopeHelper.flash(request, "Cluster has been added successfully");
+        return json(response, ResultJson.success(cluster.getId()));
+    }
+
+    private static Object deleteCluster(Request request, Response response, User user) {
+        Cluster cluster = getCluster(request);
+
+        try {
+            clusterManager.deleteCluster(cluster, user);
+            HibernateFactory.getSession().getTransaction().commit();
+        }
+        catch (Exception e) {
+            LOG.error("Deleting cluster failed", e);
+            return json(response, HttpStatus.SC_INTERNAL_SERVER_ERROR, ResultJson.error(e.getMessage()));
+        }
+        FlashScopeHelper.flash(request, String.format("Cluster '%s' deleted successfully", cluster.getName()));
```

Comment:
**This applies to all usages of the flash scope:**
These messages should be shown as "success" (green) instead of "info" (blue).

We can push `MessageResponse` objects to the scope instead. Otherwise, we have only one type of message that is interpreted as `info` on the frontend.
Instead, we should deserialize the message objects with the types, and render them with `toastr`, etc. accordingly.

Reply:
Good point. Will address this in the next iterations.